lib/locks.js

```diff
@@ -0,0 +1,162 @@
+'use strict';
+
+const { Worker, isMainThread, parentPort } = require('worker_threads');
+
+const threads = new Set();
+const resources = new Map();
+
+const LOCKED = 0;
+const UNLOCKED = 1;
+
+const sendMessage = message => {
+  if (isMainThread) {
+    for (const thread of threads) {
+      thread.worker.postMessage(message);
+    }
+  } else {
+    parentPort.postMessage(message);
+  }
+};
+
+class Lock {
+  constructor(name, options, callback) {
+    if (typeof options === 'function') {
+      callback = options;
+      options = {};
+    }
+    const { mode, ifAvailable, steal } = options;
+    this.name = name;
+    this.mode = mode || 'exclusive';
+    this.ifAvailable = ifAvailable || false;
+    this.steal = steal || false;
+    this.callback = callback;
+  }
+}
+
+class Mutex {
+  constructor(resourceName, buffer, initial = false) {
+    this.name = resourceName;
+    this.flag = new Int32Array(buffer, 0, 1);
+    if (initial) Atomics.store(this.flag, 0, UNLOCKED);
+    this.owner = false;
+    this.trying = false;
+    this.queue = [];
+    this.current = null;
+  }
+
+  enter(lock) {
+    this.queue.push(lock);
+    this.trying = true;
+    return this.tryEnter();
+  }
+
+  tryEnter() {
+    if (this.queue.length === 0) return;
+    const prev = Atomics.exchange(this.flag, 0, LOCKED);
+    if (prev === LOCKED) return;
+    this.owner = true;
+    this.trying = false;
+    const lock = this.queue.shift();
+    this.current = lock;
+    return lock.callback(lock).then(() => {
+      this.leave();
+    });
+  }
+
+  enterIfAvailable(lock) {
+    if (this.owner) return lock.callback();
+    const prev = Atomics.exchange(this.flag, 0, LOCKED);
+    if (prev === LOCKED) return lock.callback();
```

Comment:
According to https://wicg.github.io/web-locks/#api-lock-manager-request when `ifAvailable` is true the callback must be passed a `lock` (boolean) argument that signifies whether it got the lock or not.